### DISCUSSION TITLE
Problems using Netket with GPUs

### DISCUSSION THREAD

**Discussion by RomanArmenta** [netket/netket#1384] [Q&A] [answered]
Hi. I'm trying to use the GPU capabilities provided by my school with Netket, but I can't make it work somehow. I installed all the required libraries like CUDA and QDNN from the JAX repository, and when I run the script in the cluster, it looks like is running in GPU. However, even when the state of the GPUs says that 99% of the GPU memory is used by my script, the performance shown by the GPUs is worst than in the CPU of my personal computer. 

The script I'm running is the Ising chain for N=60 sites:

```python
import os
import netket as nk
from netket.operator.spin import sigmax,sigmaz 
import time

os.environ["JAX_PLATFORM_NAME"] = "gpu"
N = 60
hi = nk.hilbert.Spin(s=1 / 2, N=N)

Gamma = -1
H = sum([Gamma*sigmax(hi,i) for i in range(N)])
V=-1
H += sum([V*sigmaz(hi,i)*sigmaz(hi,(i+1)%N) for i in range(N)])

ma = nk.models.RBM(alpha=1)



# Build the sampler
sa = nk.sampler.MetropolisLocal(hi)

# Optimizer
op = nk.optimizer.Sgd(learning_rate=0.05)

# Stochastic Reconfiguration
sr = nk.optimizer.SR(diag_shift=0.1)

# The variational state
vs = nk.vqs.MCState(sa, ma, n_samples=40000)

# The ground-state optimization loop
gs = nk.VMC(
    hamiltonian=H,
    optimizer=op,
    preconditioner=sr,
    variational_state=vs)

start = time.time()
gs.run(out='RBM', n_iter=600)
end = time.time()

print('### RBM calculation')
print('Has',vs.n_parameters,'parameters')
print('The RBM calculation took',end-start,'seconds')
```

This is the performance in my PC (CPU: Intel core i5+ 8th gen)
![Captura de pantalla de 2022-11-04 15-04-25](https://user-images.githubusercontent.com/89938933/200074535-dec4e7ce-adc9-42d6-be97-7aae45ab6a76.png)

This is the performance shown by the GPUs:

![WhatsApp Image 2022-11-04 at 2 54 38 PM](https://user-images.githubusercontent.com/89938933/200074656-cae5dba5-1b4b-49c2-a6c6-8493f784d808.jpeg)

And here are the specification of the GPUs in my School's cluster (32GiB):

![WhatsApp Image 2022-11-04 at 2 54 38 PM(1)](https://user-images.githubusercontent.com/89938933/200074747-ca870ca5-11a0-4d57-a309-cb7c32db7710.jpeg)

I got some questions: Am I writing the script right to run correctly on GPUs? If is that the case, then is it normal for this performance?

I hope you can help me. Thank you so much.

**Comment by PhilipVinc** [comment, marked as answer]
Well, the volatile GPU utilisation does not tell you anything. You should properly profile your code (look into jax.profile) to identity bottlenecks.

Anyhow I can tell you:
 - you are using the default of `n_chains_per_rank=16` which is terrible for GPUs, because you will be evaluating the Markov chain with batches of 16 samples at a time. You want to have several thousands here, 
 - You are using the default of `n_discard=n_samples/10 ==> 4096` which is far too much. You should use something far lower, especially if you have many Markov chains.
 - I would do something `n_chains_per_rank=2000`, `n_discard_per_chain=20` or something like that.

Also note that you should declare `os.environ["JAX_PLATFORM_NAME"] = "gpu"` before importing netket and jax. Declaring it afterwards means it'll be ignored.

With the changes mentioned

```python
sa = nk.sampler.MetropolisLocal(hi, n_chains=2048)
...
vs = nk.vqs.MCState(sa, ma, n_samples=40000, n_discard=25)
```
above I get on a RTX 3090

```python
➜ ipython -i ppa.py
Python 3.10.6 (main, Aug 24 2022, 12:15:29) [GCC 12.1.1 20220507 (Red Hat 12.1.1-1)]
/home/filippovicentini/Dropbox/Ricerca/Codes/Python/netket/netket/vqs/mc/mc_state/state.py:60: UserWarning: n_samples=40000 (40000 per MPI rank) does not divide n_chains=2048, increased to 40960 (40960 per MPI rank)
  warnings.warn(
  2%|█▌         | 10/600 [00:16<15:53,  1.62s/it, Energy=-75.252 ± 0.012 [σ²=4.405, R̂=1.0371, τ=0.6<2.3]]]
```
which is much faster than the 6 hours your code was giving me at first.

Also, do note that unless you are using professional GPUs (like V100, A100...) your gpu is rather bad at double precision calculations and you can get a further speed up by switching to single-precision 
```python
sa = nk.sampler.MetropolisLocal(hi, n_chains=2048, dtype=jnp.float32)
ma = nk.models.RBM(alpha=1, param_dtype=jnp.float32)
vs = nk.vqs.MCState(sa, ma, n_samples=40000, n_discard=25)
```
(in this case which is memory-bandwidth bound the speedup is only a factor of ~2, but you can get up to a factor of 64 in compute-bound calculations with complicated neural networks with tons of parameters).
```python
  2%|█▌         | 10/600 [00:8<9:53,  1.62it/s, Energy=-75.447 ± 0.009 [σ²=4.805, R̂=1.0411, τ=0.5<2.2]]]
```
Do note that single precision is not our default because it does not converge sometimes and there might be numerical precision issues that we never investigated (netket does work, however, with reduced precision).

**Reply by PhilipVinc** [reply]
In general a good protocol to identify bottlenecks is to run `vs.sample()`, `vs.expect(H)`, `_, F = vs.expect_and_grad(H)` and `nk.optimizer.qgt.QGT***(vs)@F` and see what is very slow...

**Reply by PhilipVinc** [reply]
Those are the functions that are called internally by the driver. I'd say that in general any slowdown that you can do something about (other than changing hamiltonian) is in the sampling. For the other methods you don't get to tune much.
